Week2

jihyun-jeon opened the pull request:
2주차 코드리뷰 요청드립니다 :)

jihyun-jeon committed:
feat : 하샤드수,핸드폰번호가리기

jihyun-jeon committed:
chore: lodash 설치

jihyun-jeon committed:
feat : 땅따먹기

jihyun-jeon committed:
feat: 피보나치수

jihyun-jeon committed:
feat: 과일장수

jihyun-jeon committed:
feat: 튜플

jihyun-jeon commented on 땅따먹기.ts:
```diff
@@ -0,0 +1,34 @@
+function solution(land: number[][]) {
+  for (let i = 1; i < land.length; i++) {
+    for (let j = 0; j < land[i].length; j++) {
+      const prev = [...land[i - 1]]; // 비효율 개선 필요
```
배열을 계속 복사하는 로직이 비효율적인 것 같은데
개선이 필요한지 고민이 들었습니다.

f-lab-jake commented on 프로그래머스/lv1.하샤드수-구현.ts:
```diff
@@ -0,0 +1,21 @@
+
+function solution(x: number): boolean {
+  let sum = 0;
+  const strArr = String(x).split("");
+
+  strArr.forEach((s) => {
+    sum += Number(s);
+  });
```
과정을 나누세요

f-lab-jake commented on 프로그래머스/lv2.땅따먹기-DP.ts:
```diff
@@ -0,0 +1,34 @@
+ */
+
+function solution(land: number[][]) {
+  for (let i = 1; i < land.length; i++) {
+    for (let j = 0; j < land[i].length; j++) {
+      const prev = [...land[i - 1]]; // 비효율 개선 필요
+      prev[j] = 0;
+      const prevMax = Math.max(...prev);
+      land[i][j] = land[i][j] + prevMax;
+    }
+  }
```
일단 `Array.entries`를 쓰세요

f-lab-jake commented on 프로그래머스/lv2.튜플-구현.ts:
```diff
@@ -0,0 +1,31 @@
+    (a: number[], b: number[]) => a.length - b.length
+  );
+
+  arr.forEach((l: number[]) => {
+    l.forEach((n) => {
+      if (!result.includes(n)) {
+        result.push(n);
+      }
+    });
+  });
```
`flatMAp`과 `filter`를 활용해보세요

f-lab-jake commented on 프로그래머스/lv2.피보나치수-DP.ts:
```diff
@@ -0,0 +1,36 @@
+}
+
+// < Bottom-Up 방식 (반복문) > O(n)
+function solution(n: number): number {
+  let memo: number[] = [];
+
+  for (let i = 0; i <= n; i++) {
+    if (i < 2) {
+      memo[i] = i;
+    } else {
+      memo[i] = (memo[i - 2] + memo[i - 1]) % 1234567;
+    }
+  }
+
+  return memo[n];
+}
```
재귀로 다시짜보세요

jihyun-jeon committed:
refac: 2주차 코드 개선

jihyun-jeon commented:
피드백 내용 반영하여 코드 개선했습니다! [073016a](https://github.com/jihyun-jeon/ts-algorithm/pull/3/commits/073016a7782f828327d7da570287edaa06f63286)

- "땅따먹기" 문제는 Array.entries를 활용하도록 바꾸긴 했는데
    제가 의도를 잘 파악하지 못한건지, for문을 쓴 이전 로직이랑 큰 차이가 없어보이네요

f-lab-jake commented on 프로그래머스/lv2.튜플-구현.ts:
```diff
@@ -5,23 +5,19 @@
 
-  return result;
+    return l.filter((n) => !arr[idx - 1].includes(n));
+  });
```
이게 더 좋은코드 같나요? 제가 드린 제안은 코드를 더 좋게하기 위함입니다. 방향이 더 좋아지는 방향이 아닌거같으면 다시 질문을 주세요.

jihyun-jeon replied:
이 로직은 평탄화 하는 과정과 필터하는 과정이 뒤섞여 있어 좋은 방향이 아님을 파악했습니다!
이후 개선하여 이중배열을 평탄화 하는 작업과, 중복 제거 하는 로직을 분리했습니다.

f-lab-jake commented on 프로그래머스/lv1.하샤드수-구현.ts:
```diff
@@ -5,12 +5,10 @@
-  });
+  const str = String(x);
+  const strArr = str.split("");
+  const nums = strArr.map((s) => Number(s));
```
`.map(Number)`로 쓸수있습니다.

f-lab-jake replied:
이 코드가 더 낫다고 생각하나요?

jihyun-jeon replied:
숫자를 합산하는 과정에서 형변환이 이루어지는 것이 문제임을 파악했습니다!
이후 두 로직을 분리하도록 수정하였습니다.

jihyun-jeon committed:
refac: 2주차 코드 재개선
- 하샤드수 : 문자열을 숫자형으로 변환하는 작업과 숫자 더하는 로직 분리
- 튜플 : 이중배열을 순회하지 않고 배열을 평탄화한 뒤, 중복 제거하도록
- 피보나치 :  메모이제이션 함수와 재귀를 이용한 풀이

jihyun-jeon commented:
멘토링시 주신 피드백  반영하여 코드 개선했습니다! [3ad461e](https://github.com/jihyun-jeon/ts-algorithm/pull/3/commits/3ad461ee5e7e4bc13d733cc114bb4975097e89f8)
- 하샤드수 : 숫자로 형 변환 로직과 숫자 합산 로직 분리
- 튜플 : 이중배열 평탄화 작업과 중복값 필터 로직 분리 
- 피보나치 :  메모이제이션 함수와 재귀를 이용한 풀이

f-lab-jake commented on 프로그래머스/lv2.피보나치수-DP.ts:
```diff
@@ -50,6 +50,40 @@ function solution(n: number) {
+}
+
+function solution(n: number) {
+  function fibo(i: number, memoizedFn: (key: number) => number): number {
+    if (i == 0) return 0;
+    if (i == 1) return 1;
+
+    return (memoizedFn(i - 1) + memoizedFn(i - 2)) % 1234567;
+  }
```
아직 memo로직과 완전히 분리가 안됐죠?

jihyun-jeon committed:
refac: 2주차 코드 재개선
    - 튜플 : 배열 중복로직 제거를 위해 Set 활용,
    - 피보나치 :  메모 로직 분리. 메모 로직은 memoize함수에서만 처리하고 fibo는 memoize를 사용만 하도록

jihyun-jeon commented:
코드 다시 개선 해보았습니다! [e249b54](https://github.com/jihyun-jeon/ts-algorithm/pull/3/commits/e249b5402507f6414ec96da88fef006321551133)

<튜플>
기존 : filter 메서드에서 indexOf로 또 순회하며 배열의 중복값을 제거. O(n²)
개선 : Set과 has, add 메서드를 활용하여 배열의 중복 값을 제어. O(n)

<피보나치>
기존 : memoizedFibo와 fibo 함수가 얽혀 있는 상태.
개선 : memoizedFibo 함수를 분리. fibo 함수는 이제 단순히 memoizedFibo를 호출하도록

f-lab-jake commented on 프로그래머스/lv2.튜플-구현.ts:
```diff
@@ -12,7 +12,14 @@ function solution(s: any) {
-  return flatArr.filter((n: number, i: number) => flatArr.indexOf(n) === i);
+  const result = new Set<number>();
+
+  flatArr.forEach((n: number) => {
+    if (!result.has(n)) {
+      result.add(n);
+    }
+  });
```
`flatMap`은 처음에 코드를 고치기위해서 제안한 방법이었고, 지금은 필요가 없습니다.

f-lab-jake commented on 프로그래머스/lv2.피보나치수-DP.ts:
```diff
@@ -72,15 +72,15 @@ function memoize(func: Function) {
-    return (memoizedFn(i - 1) + memoizedFn(i - 2)) % 1234567;
-  }
-  const memoizedFibo = memoize(fibo);
+  return (memoizedFibo(i - 1) + memoizedFibo(i - 2)) % 1234567;
```
여전합니다 문제는. `fibo`는 자기가 memoized되던말건 신경을 안써야합니다.

jihyun-jeon committed:
refac: 2주차 코드 재수정
- 피보나치수 :fibo함수와 memo로직 분리, cache 객체 Map으로 변경
- 튜플 : flatMap이 아닌 for...of로 순회, 중복요소 확인위해 Set 활용

jihyun-jeon commented:
튜플,피보나치수 코드 재수정하였습니다!
[de3b3bc](https://github.com/jihyun-jeon/ts-algorithm/pull/3/commits/de3b3bc8c35246f2d11bd46372fafb52f4174440)
